### ISSUE TITLE
[Linux] Segfautlt on reset layout

### ISSUE THREAD

**Issue by KottV** [BespokeSynth/BespokeSynth#56]
Greetings,

It affects only linux builds. Pressing "reset layout" cause segmentation fault. GDB trace:

```
(gdb) run
Starting program: /media/src/src/audio/BespokeSynth/Builds/LinuxMakefile/build/BespokeSynth 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
JUCE v5.4.7
[Detaching after fork from child process 23923]
[Detaching after fork from child process 23924]
[New Thread 0x7ffff64ee700 (LWP 23927)]
0.001: pixel ratio: 1 screen width: 2560 screen height: 1080
[New Thread 0x7ffff578c700 (LWP 23928)]
[Thread 0x7ffff578c700 (LWP 23928) exited]
[New Thread 0x7ffff578c700 (LWP 23929)]
[New Thread 0x7ffff4f8b700 (LWP 23930)]
[New Thread 0x7fffc38d3700 (LWP 23952)]
0.001: output: Gina3G, Gina3G; Front output / input   input: Gina3G, Gina3G; Front output / input
[Thread 0x7ffff64ee700 (LWP 23927) exited]
0.547667: Loading layout: /home/kv/.config/BespokeSynth/data/layouts/blank.json
3.64367: Loading layout: /home/kv/.config/BespokeSynth/data/layouts/blank.json

Thread 1 "BespokeSynth" received signal SIGSEGV, Segmentation fault.
IDrawableModule::OnClicked (this=0x1585170, x=159, y=26, right=<optimized out>) at /usr/include/c++/10/bits/stl_vector.h:1043
1043          operator[](size_type __n) _GLIBCXX_NOEXCEPT
Missing separate debuginfos, use: zypper install krb5-debuginfo-1.18.2-3.2.x86_64 libX11-6-debuginfo-1.6.12-1.1.x86_64 libX11-xcb1-debuginfo-1.6.12-1.1.x86_64 libXau6-debuginfo-1.0.9-1.7.x86_64 libXcursor1-debuginfo-1.2.0-1.5.x86_64 libXext6-debuginfo-1.3.4-1.7.x86_64 libXfixes3-debuginfo-5.0.3-1.11.x86_64 libXinerama1-debuginfo-1.1.4-1.8.x86_64 libXrandr-devel-debuginfo-1.5.2-1.7.x86_64 libXrender1-debuginfo-0.9.10-1.12.x86_64 libXss1-debuginfo-1.2.3-1.9.x86_64 libbrotlicommon1-debuginfo-1.0.9-1.1.x86_64 libbrotlidec1-debuginfo-1.0.9-1.1.x86_64 libbz2-1-debuginfo-1.0.8-2.20.x86_64 libcom_err2-debuginfo-1.45.6-1.19.x86_64 libcrypt1-debuginfo-4.4.17-1.1.x86_64 libcurl4-debuginfo-7.72.0-1.2.x86_64 libdrm2-debuginfo-2.4.102-2.1.x86_64 libfreetype6-debuginfo-2.10.2-1.3.x86_64 libgcc_s1-debuginfo-10.2.1+git583-1.2.x86_64 libglvnd-debuginfo-1.3.2-2.1.x86_64 libidn2-0-debuginfo-2.3.0-3.2.x86_64 libkeyutils1-debuginfo-1.6-1.19.x86_64 libldap-2_4-2-debuginfo-2.4.53-57.2.x86_64 libnghttp2-14-debuginfo-1.41.0-1.3.x86_64 libopenssl1_1-debuginfo-1.1.1g-2.13.x86_64 libpcre1-debuginfo-8.44-1.19.x86_64 libpng16-16-debuginfo-1.6.37-1.7.x86_64 libpython3_8-1_0-debuginfo-3.8.5-2.1.x86_64 libsasl2-3-debuginfo-2.1.27-3.5.x86_64 libselinux1-debuginfo-3.0-2.3.x86_64 libssh4-debuginfo-0.9.5-1.1.x86_64 libstdc++6-debuginfo-10.2.1+git583-1.2.x86_64 libunistring2-debuginfo-0.9.10-2.8.x86_64 libxcb-dri3-0-debuginfo-1.14-1.3.x86_64 libxcb-glx0-debuginfo-1.14-1.3.x86_64 libxcb-present0-debuginfo-1.14-1.3.x86_64 libxcb-sync1-debuginfo-1.14-1.3.x86_64 libxcb1-debuginfo-1.14-1.3.x86_64 libz1-debuginfo-1.2.11-16.1.x86_64
(gdb) bt
#0  IDrawableModule::OnClicked (this=0x1585170, x=159, y=26, right=<optimized out>) at /usr/include/c++/10/bits/stl_vector.h:1043
#1  0x0000000000883c33 in IClickable::TestClick (testOnly=false, right=false, y=26, x=159, this=0x1585170) at ../../Source/IClickable.cpp:50
#2  IClickable::TestClick (this=0x1585170, x=159, y=26, right=<optimized out>, testOnly=<optimized out>) at ../../Source/IClickable.cpp:34
#3  0x000000000089f522 in ModularSynth::CheckClick (this=0x14505e8, clickedModule=0x1585170, x=159, y=26, rightButton=<optimized out>) at ../../Source/ModularSynth.cpp:1031
#4  0x0000000000c20d7e in juce::Component::internalMouseDown (this=<optimized out>, source=..., relativePos=..., time=..., pressure=<optimized out>, orientation=<optimized out>, rotation=<optimized out>, tiltX=<optimized out>, tiltY=<optimized out>)
    at /usr/share/juce/modules/juce_gui_basics/components/juce_Component.cpp:2407
#5  0x0000000000caf9cb in juce::MouseInputSourceInternal::sendMouseDown (time=..., screenPos=..., comp=..., this=0x143b270) at /usr/share/juce/modules/juce_gui_basics/mouse/juce_MouseInputSource.cpp:584
#6  juce::MouseInputSourceInternal::setButtons (this=this@entry=0x143b270, screenPos=..., screenPos@entry=..., time=time@entry=..., newButtonState=..., newButtonState@entry=...) at /usr/share/juce/modules/juce_gui_basics/mouse/juce_MouseInputSource.cpp:216
#7  0x0000000000c340f1 in juce::MouseInputSourceInternal::handleEvent (pen=..., newOrientation=<optimized out>, newPressure=<optimized out>, newMods=..., time=..., positionWithinPeer=..., newPeer=..., this=<optimized out>)
    at /usr/share/juce/modules/juce_gui_basics/mouse/juce_MouseInputSource.cpp:334
#8  juce::MouseInputSource::handleEvent (this=this@entry=0x7fffffffd750, peer=..., pos=..., pos@entry=..., time=time@entry=1601348466602, mods=..., mods@entry=..., pressure=<optimized out>, orientation=<optimized out>, penDetails=...)
    at /usr/share/juce/modules/juce_gui_basics/mouse/juce_MouseInputSource.cpp:634
#9  0x0000000000c342d2 in juce::ComponentPeer::handleMouseEvent (this=0x144a220, type=juce::MouseInputSource::mouse, pos=..., newMods=..., newPressure=<optimized out>, newOrientation=<optimized out>, time=1601348466602, pen=..., touchIndex=0)
    at /usr/include/c++/10/bits/atomic_base.h:325
#10 0x0000000000cb083d in juce::LinuxComponentPeer::handleButtonPressEvent (buttonPressEvent=..., this=0x144a220) at /usr/share/juce/modules/juce_gui_basics/native/juce_linux_X11_Windowing.cpp:1877
#11 juce::LinuxComponentPeer::handleWindowMessage (this=0x144a220, event=...) at /usr/share/juce/modules/juce_gui_basics/native/juce_linux_X11_Windowing.cpp:1635
#12 0x0000000000b7d48a in operator() (__closure=<optimized out>) at /usr/share/juce/modules/juce_gui_basics/native/juce_linux_X11.cpp:217
#13 std::__invoke_impl<void, juce::XWindowSystem::initialiseXDisplay()::<lambda(int)>&, int> (__f=...) at /usr/include/c++/10/bits/invoke.h:60
#14 std::__invoke_r<void, juce::XWindowSystem::initialiseXDisplay()::<lambda(int)>&, int> (__fn=...) at /usr/include/c++/10/bits/invoke.h:153
#15 std::_Function_handler<void(int), juce::XWindowSystem::initialiseXDisplay()::<lambda(int)> >::_M_invoke(const std::_Any_data &, int &&) (__functor=..., __args#0=<optimized out>) at /usr/include/c++/10/bits/std_function.h:291
#16 0x0000000000ab6c3a in std::function<void (int)>::operator()(int) const (__args#0=<optimized out>, this=0x143acb0) at /usr/include/c++/10/bits/std_function.h:617
#17 juce::InternalRunLoop::dispatchPendingEvents (this=0x143ac20) at /usr/share/juce/modules/juce_events/native/juce_linux_Messaging.cpp:169
#18 juce::MessageManager::dispatchNextMessageOnSystemQueue (returnIfNoPendingMessages=<optimized out>) at /usr/share/juce/modules/juce_events/native/juce_linux_Messaging.cpp:260
#19 0x0000000000ab6de7 in juce::MessageManager::runDispatchLoop (this=0x143abb0) at /usr/share/juce/modules/juce_events/messages/juce_MessageManager.cpp:128
#20 0x00000000004f2092 in juce::JUCEApplicationBase::main () at /usr/share/juce/modules/juce_events/messages/juce_ApplicationBase.cpp:262
#21 0x00007ffff728acca in __libc_start_main (main=0x4da4c0 <main(int, char**)>, argc=1, argv=0x7fffffffdb58, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffdb48) at ../csu/libc-start.c:308
#22 0x00000000004f4a8a in _start () at ../sysdeps/x86_64/start.S:120
```

**Comment by KottV** [comment]
I can't reproduce it under valgrind, so I guess it's maybe some race condition on closing/loading layout.

**Comment by KottV** [comment]
On third attempt crashed in valgrind --tool=helgrind:

```
==1281== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==1281==  General Protection Fault
==1281==    at 0x887983: IDrawableModule::OnClicked(int, int, bool) (IDrawableModule.cpp:551)
==1281==    by 0x883C32: TestClick (IClickable.cpp:50)
==1281==    by 0x883C32: IClickable::TestClick(int, int, bool, bool) (IClickable.cpp:34)
==1281==    by 0x89F521: ModularSynth::CheckClick(IDrawableModule*, int, int, bool) (ModularSynth.cpp:1031)
==1281==    by 0xC20D7D: juce::Component::internalMouseDown(juce::MouseInputSource, juce::Point<float>, juce::Time, float, float, float, float, float) (juce_Component.cpp:2407)
==1281==    by 0xCAF9CA: sendMouseDown (juce_MouseInputSource.cpp:143)
==1281==    by 0xCAF9CA: juce::MouseInputSourceInternal::setButtons(juce::Point<float>, juce::Time, juce::ModifierKeys) (juce_MouseInputSource.cpp:216)
==1281==    by 0xC340F0: handleEvent (juce_MouseInputSource.cpp:334)
==1281==    by 0xC340F0: juce::MouseInputSource::handleEvent(juce::ComponentPeer&, juce::Point<float>, long long, juce::ModifierKeys, float, float, juce::PenDetails const&) (juce_MouseInputSource.cpp:634)
==1281==    by 0xC342D1: juce::ComponentPeer::handleMouseEvent(juce::MouseInputSource::InputSourceType, juce::Point<float>, juce::ModifierKeys, float, float, long long, juce::PenDetails, int) (juce_ComponentPeer.cpp:88)
==1281==    by 0xCB083C: handleButtonPressEvent (juce_linux_X11_Windowing.cpp:1877)
==1281==    by 0xCB083C: juce::LinuxComponentPeer::handleWindowMessage(_XEvent&) (juce_linux_X11_Windowing.cpp:1635)
==1281==    by 0xB7D489: operator() (juce_linux_X11.cpp:217)
==1281==    by 0xB7D489: __invoke_impl<void, juce::XWindowSystem::initialiseXDisplay()::<lambda(int)>&, int> (invoke.h:60)
==1281==    by 0xB7D489: __invoke_r<void, juce::XWindowSystem::initialiseXDisplay()::<lambda(int)>&, int> (invoke.h:153)
==1281==    by 0xB7D489: std::_Function_handler<void (int), juce::XWindowSystem::initialiseXDisplay()::{lambda(int)#1}>::_M_invoke(std::_Any_data const&, int&&) (std_function.h:291)
==1281==    by 0xAB6C39: operator() (std_function.h:622)
==1281==    by 0xAB6C39: dispatchPendingEvents (juce_linux_Messaging.cpp:169)
==1281==    by 0xAB6C39: juce::MessageManager::dispatchNextMessageOnSystemQueue(bool) (juce_linux_Messaging.cpp:260)
==1281==    by 0xAB6DE6: juce::MessageManager::runDispatchLoop() (juce_MessageManager.cpp:128)
==1281==    by 0x4F2091: juce::JUCEApplicationBase::main() (juce_ApplicationBase.cpp:262)
==1281== ----------------------------------------------------------------
==1281== 
==1281== Thread #6: Exiting thread still holds 2 locks
==1281==    at 0x4C8D1B6: futex_lock_pi (futex-internal.h:412)
==1281==    by 0x4C8D1B6: __pthread_mutex_lock_full (pthread_mutex_lock.c:419)
==1281==    by 0x483DBF8: ??? (in /usr/lib64/valgrind/vgpreload_helgrind-amd64-linux.so)
==1281==    by 0x8C0D0D: lock (OpenFrameworksPort.h:137)
==1281==    by 0x8C0D0D: NamedMutex::Lock(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) (NamedMutex.cpp:18)
==1281==    by 0x8C0DFA: ScopedMutex::ScopedMutex(NamedMutex*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) (NamedMutex.cpp:38)
==1281==    by 0x8A037D: ModularSynth::AudioIn(float const**, int, int) (ModularSynth.cpp:1241)
==1281==    by 0x890E97: non-virtual thunk to MainContentComponent::audioDeviceIOCallback(float const**, int, float**, int, int) (MainComponent.cpp:136)
==1281==    by 0x92F039: juce::AudioDeviceManager::audioDeviceIOCallbackInt(float const**, int, float**, int, int) (juce_AudioDeviceManager.cpp:823)
==1281==    by 0x924483: juce::(anonymous namespace)::ALSAThread::run() (juce_linux_ALSA.cpp:716)
==1281==    by 0xA604DB: juce::Thread::threadEntryPoint() (juce_Thread.cpp:96)
==1281==    by 0xA60708: juce_threadEntryPoint (juce_Thread.cpp:118)
==1281==    by 0xA60708: juce::threadEntryProc(void*) (juce_posix_SharedCode.h:834)
==1281==    by 0x4840886: ??? (in /usr/lib64/valgrind/vgpreload_helgrind-amd64-linux.so)
==1281==    by 0x4C8AEA9: start_thread (pthread_create.c:477)
==1281== 
==1281== ----------------------------------------------------------------
==1281== 
==1281== Thread #1: Exiting thread still holds 1 lock
==1281==    at 0x887983: IDrawableModule::OnClicked(int, int, bool) (IDrawableModule.cpp:551)
==1281==    by 0x883C32: TestClick (IClickable.cpp:50)
==1281==    by 0x883C32: IClickable::TestClick(int, int, bool, bool) (IClickable.cpp:34)
==1281==    by 0x89F521: ModularSynth::CheckClick(IDrawableModule*, int, int, bool) (ModularSynth.cpp:1031)
==1281==    by 0xC20D7D: juce::Component::internalMouseDown(juce::MouseInputSource, juce::Point<float>, juce::Time, float, float, float, float, float) (juce_Component.cpp:2407)
==1281==    by 0xCAF9CA: sendMouseDown (juce_MouseInputSource.cpp:143)
==1281==    by 0xCAF9CA: juce::MouseInputSourceInternal::setButtons(juce::Point<float>, juce::Time, juce::ModifierKeys) (juce_MouseInputSource.cpp:216)
==1281==    by 0xC340F0: handleEvent (juce_MouseInputSource.cpp:334)
==1281==    by 0xC340F0: juce::MouseInputSource::handleEvent(juce::ComponentPeer&, juce::Point<float>, long long, juce::ModifierKeys, float, float, juce::PenDetails const&) (juce_MouseInputSource.cpp:634)
==1281==    by 0xC342D1: juce::ComponentPeer::handleMouseEvent(juce::MouseInputSource::InputSourceType, juce::Point<float>, juce::ModifierKeys, float, float, long long, juce::PenDetails, int) (juce_ComponentPeer.cpp:88)
==1281==    by 0xCB083C: handleButtonPressEvent (juce_linux_X11_Windowing.cpp:1877)
==1281==    by 0xCB083C: juce::LinuxComponentPeer::handleWindowMessage(_XEvent&) (juce_linux_X11_Windowing.cpp:1635)
==1281==    by 0xB7D489: operator() (juce_linux_X11.cpp:217)
==1281==    by 0xB7D489: __invoke_impl<void, juce::XWindowSystem::initialiseXDisplay()::<lambda(int)>&, int> (invoke.h:60)
==1281==    by 0xB7D489: __invoke_r<void, juce::XWindowSystem::initialiseXDisplay()::<lambda(int)>&, int> (invoke.h:153)
==1281==    by 0xB7D489: std::_Function_handler<void (int), juce::XWindowSystem::initialiseXDisplay()::{lambda(int)#1}>::_M_invoke(std::_Any_data const&, int&&) (std_function.h:291)
==1281==    by 0xAB6C39: operator() (std_function.h:622)
==1281==    by 0xAB6C39: dispatchPendingEvents (juce_linux_Messaging.cpp:169)
==1281==    by 0xAB6C39: juce::MessageManager::dispatchNextMessageOnSystemQueue(bool) (juce_linux_Messaging.cpp:260)
==1281==    by 0xAB6DE6: juce::MessageManager::runDispatchLoop() (juce_MessageManager.cpp:128)
==1281==    by 0x4F2091: juce::JUCEApplicationBase::main() (juce_ApplicationBase.cpp:262)
==1281== 
==1281== 
==1281== Use --history-level=approx or =none to gain increased speed, at
==1281== the cost of reduced accuracy of conflicting-access information
==1281== For lists of detected and suppressed errors, rerun with: -s
==1281== ERROR SUMMARY: 1614079 errors from 129 contexts (suppressed: 397953 from 963)
```

**Comment by awwbees** [comment]
thanks! that output was helpful. I think it's just dumb luck that it wasn't crashing on other platforms. fixed in dc5b6094e594c5c2a4d6543fb3d967f7b1aac3ba
